lambdas/pkgpush/index.py

```diff
@@ -172,6 +179,51 @@
 logger = get_quilt_logger()
 
 
+def calculate_pkg_hashes(s3_client, pkg):
+    entries = []
+    for lk, entry in pkg.walk():
+        if entry.hash is not None:
+            continue
+        if entry.size > S3_HASH_LAMBDA_MAX_FILE_SIZE:
+            raise FileTooLargeForHashing(lk)
+
+        entries.append(entry)
+
+    with concurrent.futures.ThreadPoolExecutor(max_workers=S3_HASH_LAMBDA_CONCURRENCY) as pool:
+        fs = [
+            pool.submit(calculate_pkg_entry_hash, s3_client, entry)
+            for lk, entry in pkg.walk()
+            if entry.hash is None
+        ]
+        for f in concurrent.futures.as_completed(fs):
+            f.result()
+
+
+def calculate_pkg_entry_hash(s3_client, pkg_entry):
+    pk = pkg_entry.physical_key
+    params = {
+        'Bucket': pk.bucket,
+        'Key': pk.path,
+    }
+    if pk.version_id is not None:
+        params['VersionId'] = pk.version_id
+    url = s3_client.generate_presigned_url(
+        ClientMethod='get_object',
+        ExpiresIn=S3_HASH_LAMBDA_SIGNED_URL_EXPIRES_IN,
+        Params=params,
+    )
+    pkg_entry.hash = {
+        'type': 'SHA256',
+        'value': invoke_hash_lambda(url),
+    }
+
+
+def invoke_hash_lambda(url):
+    resp = lambda_.invoke(FunctionName=S3_HASH_LAMBDA, Payload=json.dumps(url))
+    assert 'FunctionError' not in resp
```

Comment:
Shouldn't we be checking status code as well? What happens if this assert fails to concurrent hashes in progress?

Reply:
> Shouldn't we be checking status code as well?

No, I think we get boto3 exceptions in this case.

Reply:
> What happens if this assert fails to concurrent hashes in progress?

Actually that's a good question, unfortunately we can't "cancel" already invoked lambdas.